### ISSUE TITLE
Error while building

### ISSUE THREAD

**Issue by xysDavid** [cloudflare/workerd#652]
```
❯ bazel build //src/workerd/server:workerd
INFO: Analyzed target //src/workerd/server:workerd (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /Users/davidysxue/dev/git/workerd/src/workerd/jsg/BUILD.bazel:7:14: Compiling src/workerd/jsg/ser.c++ failed: (Exit 1): cc_wrapper.sh failed: error executing command (from target //src/workerd/jsg:jsg)
  (cd /private/var/tmp/_bazel_davidysxue/6eed07a0cee52db6c2f580dc80a0d858/sandbox/darwin-sandbox/4999/execroot/workerd && \
  exec env - \
    BAZEL_COMPILER=clang \
    CC=clang \
    CXX=clang++ \
    PATH=/usr/local/anaconda3/bin:/usr/local/anaconda3/condabin:/Users/davidysxue/.pyenv/shims:/Users/davidysxue/Library/pnpm:/Users/davidysxue/.nvm/versions/node/v14.21.3/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/Wireshark.app/Contents/MacOS:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/davidysxue/.cargo/bin \
    PWD=/proc/self/cwd \
  external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++0x' -MD -MF bazel-out/darwin-fastbuild/bin/src/workerd/jsg/_objs/jsg/ser.pic.d '-frandom-seed=bazel-out/darwin-fastbuild/bin/src/workerd/jsg/_objs/jsg/ser.pic.o' -fPIC -DKJ_HAS_OPENSSL -DKJ_HAS_ZLIB -DKJ_HAS_LIBDL '-DKJ_SAVE_ACQUIRED_LOCK_INFO=0' '-DKJ_TRACK_LOCK_BLOCKING=0' '-DV8_TYPED_ARRAY_MAX_SIZE_IN_HEAP=64' -DV8_DEPRECATION_WARNINGS -DV8_IMMINENT_DEPRECATION_WARNINGS -DV8_ENABLE_LAZY_SOURCE_POSITIONS -DV8_ENABLE_TURBOFAN -DV8_ENABLE_WEBASSEMBLY -DGOOGLE3 -DV8_ADVANCED_BIGINT_ALGORITHMS -DV8_CONCURRENT_MARKING -DV8_TARGET_ARCH_X64 -DV8_HAVE_TARGET_OS -DV8_TARGET_OS_MACOS -DV8_COMPRESS_POINTERS -DV8_31BIT_SMIS_ON_64BIT_ARCH -DV8_COMPRESS_POINTERS_IN_SHARED_CAGE -DV8_SHORT_BUILTIN_CALLS -DU_COMMON_IMPLEMENTATION -DU_ICUDATAENTRY_IN_COMMON '-DHAVE_DLOPEN=0' '-DUCONFIG_ONLY_HTML_CONVERSION=1' '-DU_CHARSET_IS_UTF8=1' '-DU_USING_ICU_NAMESPACE=0' '-DU_ENABLE_DYLOAD=0' '-DUSE_CHROMIUM_ICU=1' '-DU_ENABLE_TRACING=1' '-DU_ENABLE_RESOURCE_TRACING=0' '-DUNISTR_FROM_STRING_EXPLICIT=' '-DUNISTR_FROM_CHAR_EXPLICIT=' -DU_I18N_IMPLEMENTATION -DWORKERD_ICU_DATA_EMBED '-DBAZEL_CURRENT_REPOSITORY=""' -iquote . -iquote bazel-out/darwin-fastbuild/bin -iquote external/capnp-cpp -iquote bazel-out/darwin-fastbuild/bin/external/capnp-cpp -iquote external/zlib -iquote bazel-out/darwin-fastbuild/bin/external/zlib -iquote external/ssl -iquote bazel-out/darwin-fastbuild/bin/external/ssl -iquote external/workerd-v8 -iquote bazel-out/darwin-fastbuild/bin/external/workerd-v8 -iquote external/v8 -iquote bazel-out/darwin-fastbuild/bin/external/v8 -iquote external/com_googlesource_chromium_base_trace_event_common -iquote bazel-out/darwin-fastbuild/bin/external/com_googlesource_chromium_base_trace_event_common -iquote external/com_googlesource_chromium_icu -iquote bazel-out/darwin-fastbuild/bin/external/com_googlesource_chromium_icu -Ibazel-out/darwin-fastbuild/bin/src/workerd/jsg/_virtual_includes/jsg -Ibazel-out/darwin-fastbuild/bin/src/workerd/jsg/_virtual_includes/exception -Ibazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/kj/_virtual_includes/kj -Ibazel-out/darwin-fastbuild/bin/src/workerd/jsg/_virtual_includes/modules_capnp -Ibazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/capnp/_virtual_includes/capnp -Ibazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/kj/_virtual_includes/kj-async -Ibazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/capnp/compat/_virtual_includes/json -Ibazel-out/darwin-fastbuild/bin/src/workerd/util/_virtual_includes/util -Ibazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/kj/compat/_virtual_includes/kj-http -Ibazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/kj/compat/_virtual_includes/kj-tls -Ibazel-out/darwin-fastbuild/bin/src/workerd/util/_virtual_includes/sentry -Ibazel-out/darwin-fastbuild/bin/src/workerd/util/_virtual_includes/thread-scopes -Ibazel-out/darwin-fastbuild/bin/external/com_googlesource_chromium_base_trace_event_common/_virtual_includes/trace_event_common -Ibazel-out/darwin-fastbuild/bin/external/v8/_virtual_includes/icu/generated_torque_definitions_headers -Ibazel-out/darwin-fastbuild/bin/external/com_googlesource_chromium_icu/_virtual_includes/icu -Ibazel-out/darwin-fastbuild/bin/_virtual_includes/icudata-embed -isystem external/zlib -isystem bazel-out/darwin-fastbuild/bin/external/zlib -isystem external/ssl/src/include -isystem bazel-out/darwin-fastbuild/bin/external/ssl/src/include -isystem external/v8/icu -isystem bazel-out/darwin-fastbuild/bin/external/v8/icu -isystem external/v8/include -isystem bazel-out/darwin-fastbuild/bin/external/v8/include -isystem external/com_googlesource_chromium_icu/source/common -isystem bazel-out/darwin-fastbuild/bin/external/com_googlesource_chromium_icu/source/common -isystem external/com_googlesource_chromium_icu/source/i18n -isystem bazel-out/darwin-fastbuild/bin/external/com_googlesource_chromium_icu/source/i18n '-std=c++20' -fcoroutines-ts '-stdlib=libc++' -Wall -Wextra -Wunused-function -Wunused-lambda-capture -Wunused-variable -Wno-strict-aliasing -Wno-sign-compare -Wno-unused-parameter -Wno-missing-field-initializers -Wno-ignored-qualifiers -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c src/workerd/jsg/ser.c++ -o bazel-out/darwin-fastbuild/bin/src/workerd/jsg/_objs/jsg/ser.pic.o)
# Configuration: a700fc99f9a6974523b79a5a9d1d867b27469e9d4ddce1138d9bb76b52daa2d5
# Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/workerd/jsg/ser.c++:15:30: error: no member named 'SetWriteVersion' in 'v8::ValueSerializer'
    KJ_ASSERT(jsg::check(ser.SetWriteVersion(*version)));
                         ~~~ ^
bazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/debug.h:343:19: note: expanded from macro 'KJ_ASSERT'
#define KJ_ASSERT KJ_REQUIRE
                  ^
bazel-out/darwin-fastbuild/bin/external/capnp-cpp/src/kj/_virtual_includes/kj/kj/debug.h:228:52: note: expanded from macro 'KJ_REQUIRE'
  if (auto _kjCondition = ::kj::_::MAGIC_ASSERT << cond) {} else \
                                                   ^~~~
src/workerd/jsg/ser.c++:85:11: error: no member named 'SetWireFormatVersion' in 'v8::ValueDeserializer'
    deser.SetWireFormatVersion(*version);
    ~~~~~ ^
2 errors generated.
Target //src/workerd/server:workerd failed to build
INFO: Elapsed time: 2.902s, Critical Path: 2.42s
INFO: 17 processes: 16 internal, 1 darwin-sandbox.
FAILED: Build did NOT complete successfully
```

**Comment by xysDavid** [comment]
> Thanks for the report.
> 
> This failure doesn't repro here with ToT here ([ae92f58](https://github.com/cloudflare/workerd/commit/ae92f58e71776648cd1cbc780a9b3aabf73d108c)) on OS X.
> 
> `SetWireFormatVersion` comes from `workerd/patches/v8/patches/v8/0001-Allow-manually-setting-ValueDeserializer-format-vers.patch`. It looks like your build doesn't have the patch applied, and I'm not sure how that would happen.
> 
> ```
> $ cd workerd
> $ grep -r SetWireFormat external/v8
> external/v8/include/v8-value-serializer.h:  void SetWireFormatVersion(uint32_t version);
> external/v8/src/objects/value-serializer.h:  void SetWireFormatVersion(uint32_t version) { version_ = version; }
> external/v8/src/api/api.cc:void ValueDeserializer::SetWireFormatVersion(uint32_t version) {
> external/v8/src/api/api.cc:  private_->deserializer.SetWireFormatVersion(version);
> ```
> 
> The patches applied to v8 are listed in the workerd `WORKSPACE` file.

I think the patches are not applied in my building, too. How can I manually apply the patches ?

And I could find `SetWireFormatVersion` in `workerd/patches/v8/patches/v8/0001-Allow-manually-setting-ValueDeserializer-format-vers.patch`. But no `external/v8` is found under `workerd`, you mean `bazel-workerd`? 

```
$ cd bazel-workerd/external/v8
$ grep -r SetWireFormat *
include/v8-value-serializer.h:  void SetWireFormatVersion(uint32_t version);
src/objects/value-serializer.h:  void SetWireFormatVersion(uint32_t version) { version_ = version; }
src/api/api.cc:void ValueDeserializer::SetWireFormatVersion(uint32_t version) {
src/api/api.cc:  private_->deserializer.SetWireFormatVersion(version);
```

And sometimes I report errors like this:

```
# Configuration: a700fc99f9a6974523b79a5a9d1d867b27469e9d4ddce1138d9bb76b52daa2d5
# Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/workerd/api/encoding.c++:486:39: error: no member named 'MaybeNew' in 'v8::ArrayBuffer'
  auto maybeBuffer = v8::ArrayBuffer::MaybeNew(isolate, str->Utf8Length(isolate));
                     ~~~~~~~~~~~~~~~~~^
1 error generated.
Target //src/workerd/server:workerd failed to build
INFO: Elapsed time: 61.559s, Critical Path: 5.50s
INFO: 523 processes: 35 internal, 488 darwin-sandbox.
FAILED: Build did NOT complete successfully
```

I think these errors are causing by the same reason that patches not applied in building progress.

**Comment by ohodson** [comment]
Yes, the other errors you have come from the same issue (`MaybeNew`).

The patches can be applied manually from `workerd/patches/v8/` to v8 using `git apply` taking care to apply them in ascending order.

The patches should have applied as part of the build, but once the v8 repo has been cloned by bazel it won't try again until the v8 version workerd depends upon changes. If you are not using your own v8 customizations, does `bazel clean --expunge` followed by `bazel build //src/workerd/server:workerd` fix the issue?

---
TL;DR apologies for the confusion about `external/`, that is symlink present locally that exists as a side-effect of generating `compile_commands.json`. There is a note in [docs/vscode](https://github.com/cloudflare/workerd/blob/main/docs/vscode.md#generating-compile_commandsjson) about that.

**Comment by xysDavid** [comment]
I've tried `bazel clean --expunge` before `bazel build //src/workerd/server:workerd`. But the same error comes out again.

**Comment by ohodson** [comment]
Have you ever been able to build workerd on this machine, or is this a first time failure? Is it an Intel or ARM OS X machine?

The config for the patches is in the [WORKSPACE](https://github.com/cloudflare/workerd/blob/665681df96b97129b7008ef071ecf9ff86f78faf/WORKSPACE#L263) file.

Bazel usually uses its own patch code, but you try using an external tool there, e.g. add `patch_tool = "/usr/bin/patch",` and `verbose = True`, and then running `bazel sync` before trying `bazel build //src/workerd/server:workerd`.

You might also try using a non-existent external patch tool to check that Bazel is attempting to apply patches in the first place, e.g. `patch_tool = "notpatch"` which should then fail during patching, so you can at least see that it is happening at all.
